cpp/src/copying/contiguous_split.cu

```diff
@@ -14,466 +14,1129 @@
+ * For each input column to be split we need to know several pieces of information
+ * in the copy kernel.  This function traverses the input columns and prepares this
+ * information for the gpu.
  *
- * The table_view contained within the contiguous_split_result will pass an expect_tables_equal()
- * call with the input table.  The memory referenced by the table_view and its internal column_views
- * is entirely contained in single block of memory.
+ * This function is called recursively in the case of nested types.
+ *
+ * @param begin Beginning of input columns
+ * @param end End of input columns
+ * @param head Beginning of source buffer info array
+ * @param output Pair containing current source buffer info pointer and an integer
+ * representing our current offset nesting depth (how many list levels deep we are)
+ * @param parent_offset_index Index into src_buf_info output array indicating our nearest
+ * containing list parent. -1 if we have no list parent
+ * @param offset_depth Current offset nesting depth (how many list levels deep we are)
+ *
+ * @returns new src_buf_output after processing this range of input columns
  */
-contiguous_split_result alloc_and_copy(cudf::table_view const& t,
-                                       rmm::device_vector<column_split_info>& device_split_info,
-                                       rmm::mr::device_memory_resource* mr,
-                                       cudaStream_t stream)
-{
-  // preprocess column split information for string columns.
-  thrust::host_vector<column_split_info> split_info =
-    preprocess_string_column_info(t, device_split_info, stream);
+template <typename InputIter>
+src_buf_output setup_source_buf_info(InputIter begin,
+                                     InputIter end,
+                                     src_buf_info* head,
+                                     src_buf_output output,
+                                     int parent_offset_index = -1,
+                                     int offset_depth        = 0);
 
-  // compute the rest of the column sizes (non-string columns, and total buffer size)
-  size_t total_size      = 0;
-  size_type column_index = 0;
-  std::for_each(
-    t.begin(), t.end(), [&total_size, &column_index, &split_info](cudf::column_view const& c) {
-      total_size +=
-        cudf::type_dispatcher(c.type(), column_buffer_size_functor{}, c, split_info[column_index]);
-      column_index++;
-    });
+/**
+ * @brief Functor that builds source buffer information based on input columns.
+ *
+ * Called by setup_source_buf_info to build info for a single source column.  This function will
+ * recursively call setup_source_buf_info in the case of nested types.
+ */
+struct buf_info_functor {
+  template <typename T, std::enable_if_t<std::is_same<T, cudf::string_view>::value>* = nullptr>
+  src_buf_output operator()(column_view const& col,
+                            src_buf_info* head,
+                            src_buf_output output,
```

Comment:
The way `output` is used makes this code hard to follow. It's effectively being used as an input/output parameter even though it is passed by value. It should be more clear how the object returned from this function depends on the input `output` object (`output` also seems like a misleading name).